packages/forms/src/UIForm/fieldsets/Array/Array.component.js

```diff
@@ -0,0 +1,186 @@
+import React, { PropTypes } from 'react';
+import ArrayItem from './ArrayItem.component';
+import Message from '../../Message';
+import Widget from '../../Widget';
+import { shiftArrayErrorsKeys } from '../../utils/validation';
+
+import theme from './Array.scss';
+
+function adaptKeyWithIndex(keys, index) {
+	let indexedKeys = keys;
+	const firstIndexPlaceholder = indexedKeys.indexOf('');
+	if (firstIndexPlaceholder >= 0) {
+		indexedKeys = [...keys];
+		indexedKeys[firstIndexPlaceholder] = index;
+	}
+	return indexedKeys;
+}
+
+function getRange(previousIndex, nextIndex) {
+	const range = {};
+	if (previousIndex < nextIndex) {
+		range.minIndex = previousIndex;
+		range.maxIndex = nextIndex + 1;
+	} else {
+		range.minIndex = nextIndex;
+		range.maxIndex = previousIndex + 1;
+	}
+	return range;
+}
+
+function getItemSchema(arraySchema, index) {
+	// insert index in all fields
+	const items = arraySchema.items.map(item => ({
+		...item,
+		key: adaptKeyWithIndex(item.key, index),
+	}));
+
+	// insert index in item schema key
+	const key = arraySchema.key.concat(index);
+
+	return {
+		key,
+		items,
+		widget: arraySchema.itemWidget || 'fieldset',
+	};
+}
+
+export default class ArrayWidget extends React.Component {
+	constructor(props) {
+		super(props);
+
+		this.onAdd = this.onAdd.bind(this);
+		this.onRemove = this.onRemove.bind(this);
+		this.onReorder = this.onReorder.bind(this);
+	}
+
+	onAdd(event) {
+		const schema = this.props.schema;
+		const defaultValue = schema.schema.items.type === 'object' ? {} : '';
```

Comment:
Yes, `const schema = this.props.schema;` refers the the array schema from json-schema-form-core. But in it we have a property `schema` with the original schema value from json input.

I'll change the names to be a little clearer.
```
const arrayMergedSchema = this.props.schema;
const defaultValue = arrayMergedSchema.schema. ...
```